pkg/importer/spanner/sysl.arrai

```diff
@@ -0,0 +1,128 @@
+        attr.options:
+            let opt = //seq.split('=',attr.options);
+            $`${opt(0)}='${opt(1)}'`,
+        };
+    let pk = cond {
+        (entity.primary_key where //seq.contains(attr.name, .)):
+            '~pk' ++ sortingOrder(entity, attr)
+    };
+    let fk = cond {
+        (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .)) where .:
+            '~fk'
+    };
+    let length = cond {
+        attr.length = 'MAX': '~max',
+    };
+    let hexPrefix = cond {
+        //seq.has_prefix('0x', attr.length): '~hex',
+    };
+    //seq.join(', ', ([options, pk, fk, length, hexPrefix] where .@item != ''));
+
+# entityPatterns generates the patterns for an entity.
+let entityPatterns = \entity \model
+    let pk = cond {
+        entity.primary_key count > 1 && compareColumnOrder(entity):
+            $`primary_key="${entity.primary_key orderby .::, }"`,
+    };
+    let cluster = cond {
+        entity.cluster:
+            //seq.join('', entity.cluster >>
+            $`interleave_in_parent="${.interleaved_in}", interleave_on_delete="${.on_delete}"`),
+    };
+    let fk = cond {
+        entity.foreign_keys:
+            $`foreign_keys=[${//seq.join(', ',(entity.foreign_keys =>
+                \keys $`["constraint:${keys.constraint_name}","columns:${//seq.join(', ', (keys.foreign_keys => .attribute) orderby .)}"]`) orderby .)}]`,
+    };
+    let indices = cond {
+        (model.indexes where .table_name = entity.name):
+            $`indexes=[${//seq.join(', ', ((model.indexes orderby .) where .@item.table_name = entity.name) >>
+                $`"name:${.table_name}","unique:${.unique}","null_filtered:${.nullfiltered}","key_parts:${//seq.join(', ', .key_part orderby .)}${cond {.storing_col: $`","storing:${//seq.join(', ',.storing_col)}`}}${cond {.interleaved_table: $`","interleave_in:${.interleaved_table}`}}`)}"]`,
+    };
+    //seq.join(', ', ([pk, cluster, fk, indices] where .@item != ''));
+
+# appendEntityPatterns appends the entity patterns together.
+let appendEntityPatterns = \entity \model
+    cond {
+        hasEntityPattern(entity, model): $`[${entityPatterns(entity, model)}]`
+    };
+
+# appendAttributePatterns appends the attribute patterns together.
+let appendAttributePatterns = \entity \attr
+    cond {
+        hasAttributePattern(entity,attr): $`[${attributePatterns(entity, attr)}]`
+    };
+
+# typeInfo generates the type info for an attribute.
+let typeInfo = \entity \attr \type \isArray
+    let fkAttr = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .) where .;
+    let fkTable = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_table orderby .) where .;
+    cond {
+        fkAttr | fkTable:  //seq.join('', (fkTable => //seq.join('', .) orderby .)) ++ '.'
+            ++ //seq.join('', (fkAttr => //seq.join('', .) orderby .)),
```

Comment:
Upon closer inspection, this seems to require that both `fkAttr` and `fkTable` are non-empty? In that case `fkAttr && fkTable` would be a better condition.

Also the `//seq.join('', ...)` calls are adding a lot of noise. Perhaps `flatJoin` wasn't the best idea, but it would still be nice to get rid of them and focus more on the data you're operating on. Maybe:
```suggestion
        fkAttr && fkTable: $`${fkTable => $`${.::}` orderby .::}.${fkAttr => $`${.::}` orderby .::}`,
```

Or if you really don't like the `${.::}` pattern, consider using currying to do something like `let join = //seq.join('')`, with which you could just do a simpler:

```
$`${join(fkTable => join(.) orderby .)}.${join(fkAttr => join(.) orderby .)}`
```